Add docker config slicing support

jamiesprax opened the pull request:
This add support to slice docker configs across multiple projects.
Supported configs are:
* Dockerfile Path
* Repo Name
* Repo Tags
* Registry URL

Commit:
Add docker config slicing support

This add support to slice configs across multiple projects.
Supported configs are:
* Dockerfile Path
* Repo Name
* Repo Tags
* Registry URL

AdamDM commented on src/main/java/configurationslicing/docker/AbstractDockerSlicerSpec.java:
```diff
@@ -0,0 +1,177 @@
+    public List<AbstractProject<?, ?>> getWorkDomain() {
+        List<AbstractProject<?, ?>> list = new ArrayList<AbstractProject<?, ?>>();
+        List<AbstractProject> temp = Jenkins.getInstance().getAllItems(AbstractProject.class);
+        for (AbstractProject p: temp) {
```
don't call variables 'p'

Dalamar42 replied:
`list`, `temp` and `p` are not very descriptive. They don't explain what each of them do so it's hard to understand what the method's purpose is.

AdamDM commented on src/main/java/configurationslicing/docker/DockerRepoTagSlicer.java:
```diff
@@ -0,0 +1,39 @@
+
+        @Override
+        public String getName() {
+            return "Docker Slicer - Repo Tags";
```
magic string

AdamDM commented:
Small style comments, OLGTM

Commit:
Fix bug causing DockerBuilder to be removed from project

Due to handling of setting empty fields, setting them would cause
the whole DockerBuilder to be removed as an empty field was
essentially invalidating the DockerBuilder.

Null and empty cases are now handled for all supported configs.

Dalamar42 commented on src/main/java/configurationslicing/docker/AbstractDockerSlicerSpec.java:
```diff
@@ -0,0 +1,185 @@
+            }
+        }
+
+        // add any new ones (can't see when this would happen)
```
About this two comments. These scenarios are not supposed to happen because we don't need them or because they should be impossible?

jamiesprax replied:
Removing any new ones should now be impossible due to how i'm handling values so I could remove that one? 
Adding a new one could still be possible by adding a repo name to a project that was not initially set up with Docker Build & Publish as a build step. You don't want to do this though as you will only be able to set the sliceable configs after that and theres a bunch that we don't slice that should be set.

Dalamar42 replied:
From what you are saying I think you can omit the first one, but throw some kind of error on the second one? Is there a standard mechanism for the plugin to fail gracefully?

jamiesprax replied:
I didn't find anything that could display an error as part of the page so you'd know instantly but it would be in the logs.

I've decided to remove the delete builder part because i do newBuilders[i] = oldBuilders[i] and theres no situation which would set newBuilder[i] to null which makes the delete check obsolete. The adding builder error message is now more verbose.

Dalamar42 reviewed (approved):
Conditionally LGMT

I can't comment on code style, because I don't know what this project's code style is.

In terms of functionality I'm also not very familiar with the logic, but this looks reasonable and well tested.

emilssolmanis commented on src/main/java/configurationslicing/docker/AbstractDockerSlicerSpec.java:
```diff
@@ -0,0 +1,185 @@
+
+    private static final Logger log = LoggerFactory.getLogger(AbstractDockerSlicerSpec.class);
+
+    static final String NOTHING = "(Nothing)";
```
Why package private?

jamiesprax replied:
IJ's advice. Should I just have it public?

Dalamar42 replied:
Generally we only use package private to make things visible to their tests. If that's what this is then I think it's fine

emilssolmanis reviewed (approved):
lgtm, minor nitpicks

Commit:
Shuffle logging & make error's more informative

Also removed impossible removeBuilder condition
